compiler-access/src/main/scala/org/alephium/ralph/lsp/access/compiler/ast/node/NodeBuilder.scala

```diff
@@ -0,0 +1,90 @@
+package org.alephium.ralph.lsp.access.compiler.ast.node
+
+import com.typesafe.scalalogging.StrictLogging
+import org.alephium.ralph.Ast.Positioned
+import org.alephium.ralph.{Ast, SourceIndex}
+
+/** Functions that build a traversable tree from [[Ast.ContractWithState]], returning the root [[Node]]. */
+object NodeBuilder extends StrictLogging {
+
+  /**
+   * Given an [[Ast.ContractWithState]], builds a traversable tree.
+   *
+   * @param ast The [[Ast.ContractWithState]] instance
+   * @return Root node of the tree.
+   */
+  def buildRootNode(ast: Ast.ContractWithState,
+                    rootIndex: SourceIndex): Node[Positioned] = {
+    // TODO: Are all these siblings? If they are not, they need to build a tree structure using source-index.
+    val rootSiblings =
+      ast match {
+        case ast: Ast.TxScript =>
+          buildOne(ast.ident) ++
+            buildMany(ast.templateVars) ++
+            buildMany(ast.funcs)
+
+        case ast: Ast.Contract =>
+          buildOne(ast.stdInterfaceId) ++
+            buildOne(ast.ident) ++
+            buildMany(ast.templateVars) ++
+            buildMany(ast.fields) ++
+            buildMany(ast.funcs) ++
+            buildMany(ast.events) ++
+            buildMany(ast.constantVars) ++
+            buildMany(ast.enums) ++
+            buildMany(ast.inheritances)
+
+        case ast: Ast.ContractInterface =>
+          buildOne(ast.stdId) ++
+            buildOne(ast.ident) ++
+            buildMany(ast.funcs) ++
+            buildMany(ast.events) ++
+            buildMany(ast.inheritances)
+      }
+
+    // sort the sibling according to their source-index i.e. following their order of position in code.
+    val sortedRootSiblings =
+      rootSiblings.sortBy(_.data.sourceIndex.map(_.index))
```

Comment:
will it still be relevant when we'll have the `fileURI` as well in the `sourceIndex`?

Reply:
Yes it will. `fileURI` does not make any difference here.